build/driver.go

```diff
@@ -1,305 +0,0 @@
-	if len(r.nodes) == 0 {
-		return nil, nil
-	}
-
-	nodes := map[string][]*resolvedNode{}
-	for k, opt := range opt {
-		node, perfect, err := r.resolve(ctx, opt.Platforms, pw, platforms.OnlyStrict, nil)
-		if err != nil {
-			return nil, err
-		}
-		if !perfect {
-			break
-		}
-		nodes[k] = node
-	}
-	if len(nodes) != len(opt) {
-		// if we didn't get a perfect match, we need to boot all drivers
-		allIndexes := make([]int, len(r.nodes))
-		for i := range allIndexes {
-			allIndexes[i] = i
-		}
-
-		clients, err := r.boot(ctx, allIndexes, pw)
-		if err != nil {
-			return nil, err
-		}
-		eg, egCtx := errgroup.WithContext(ctx)
-		workers := make([][]specs.Platform, len(clients))
-		for i, c := range clients {
-			i, c := i, c
-			if c == nil {
-				continue
-			}
-			eg.Go(func() error {
-				ww, err := c.ListWorkers(egCtx)
-				if err != nil {
-					return errors.Wrap(err, "listing workers")
-				}
-
-				ps := make(map[string]specs.Platform, len(ww))
-				for _, w := range ww {
-					for _, p := range w.Platforms {
-						pk := platforms.Format(platforms.Normalize(p))
-						ps[pk] = p
-					}
-				}
-				for _, p := range ps {
-					workers[i] = append(workers[i], p)
-				}
-				return nil
-			})
-		}
-		if err := eg.Wait(); err != nil {
-			return nil, err
-		}
-
-		// then we can attempt to match against all the available platforms
-		// (this time we don't care about imperfect matches)
-		nodes = map[string][]*resolvedNode{}
-		for k, opt := range opt {
-			node, _, err := r.resolve(ctx, opt.Platforms, pw, platforms.Only, func(idx int, n builder.Node) []specs.Platform {
-				return workers[idx]
-			})
-			if err != nil {
-				return nil, err
-			}
-			nodes[k] = node
-		}
-	}
-
-	idxs := make([]int, 0, len(r.nodes))
-	for _, nodes := range nodes {
-		for _, node := range nodes {
-			idxs = append(idxs, node.driverIndex)
-		}
-	}
-
-	// preload capabilities
-	span, ctx := tracing.StartSpan(ctx, "load buildkit capabilities", trace.WithSpanKind(trace.SpanKindInternal))
-	_, err := r.opts(ctx, idxs, pw)
-	tracing.FinishWithError(span, err)
-	if err != nil {
-		return nil, err
-	}
-
-	return nodes, nil
-}
-
-func (r *nodeResolver) resolve(ctx context.Context, ps []specs.Platform, pw progress.Writer, matcher matchMaker, additional func(idx int, n builder.Node) []specs.Platform) ([]*resolvedNode, bool, error) {
-	if len(r.nodes) == 0 {
-		return nil, true, nil
-	}
-
-	if len(ps) == 0 {
-		ps = []specs.Platform{platforms.DefaultSpec()}
-	}
-
-	perfect := true
-	nodeIdxs := make([]int, 0)
-	for _, p := range ps {
-		idx := r.get(p, matcher, additional)
-		if idx == -1 {
-			idx = 0
-			perfect = false
-		}
-		nodeIdxs = append(nodeIdxs, idx)
-	}
-
-	var nodes []*resolvedNode
-	for i, idx := range nodeIdxs {
-		nodes = append(nodes, &resolvedNode{
-			resolver:    r,
-			driverIndex: idx,
-			platforms:   []specs.Platform{ps[i]},
```

Comment:
I think this is the line causing the DefaultSpec issue. It should not be set here if the default case with no specified platforms from above was bit.